File: source/SharpFlame/clsBooleanMap.cs
using Microsoft.VisualBasic;
using SharpFlame.Maths;

namespace SharpFlame
{
    public class clsBooleanMap
    {
        public class clsValueData
        {
            public sXY_int Size;
            public bool[,] Value;
        }

        public clsValueData ValueData = new clsValueData();

        public void Blank(int SizeX, int SizeY)
        {
            ValueData.Size.X = SizeX;
            ValueData.Size.Y = SizeY;
            ValueData.Value = new bool[SizeY, SizeX];
        }

        public void SizeCopy(clsBooleanMap Source)
        {
            ValueData.Size.X = Source.ValueData.Size.X;
            ValueData.Size.Y = Source.ValueData.Size.Y;
            ValueData.Value = new bool[ValueData.Size.Y, ValueData.Size.X];
        }

        public void Copy(clsBooleanMap Source)
        {
            int X = 0;
            int Y = 0;

            SizeCopy(Source);
            for ( Y = 0; Y <= Source.ValueData.Size.Y - 1; Y++ )
            {
                for ( X = 0; X <= Source.ValueData.Size.X - 1; X++ )
                {
                    ValueData.Value[Y, X] = Source.ValueData.Value[Y, X];
                }
            }
        }

        public void Convert_Heightmap(clsHeightmap Source, long AtOrAboveThisHeightEqualsTrue)
        {
            int X = 0;
            int Y = 0;

            ValueData.Size.X = Source.HeightData.SizeX;
            ValueData.Size.Y = Source.HeightData.SizeY;
            ValueData.Value = new bool[ValueData.Size.Y, ValueData.Size.X];
            for ( Y = 0; Y <= Source.HeightData.SizeY - 1; Y++ )
            {
                for ( X = 0; X <= Source.HeightData.SizeX - 1; X++ )
                {
                    ValueData.Value[Y, X] = Source.HeightData.Height[Y, X] >= AtOrAboveThisHeightEqualsTrue;
                }
            }
        }

        public void Remove_Diagonals()
        {
            int X = 0;
            int Y = 0;
            bool Flag = default(bool);

            X = 0;
            Y = 0;
            while ( Y < ValueData.Size.Y - 1 )
            {
                X = 0;
                while ( X < ValueData.Size.X - 1 )
                {
                    Flag = false;
                    if ( ValueData.Value[Y, X] )
                    {
                        if ( ValueData.Value[Y, X + 1] )
                        {
                            if ( ValueData.Value[Y + 1, X] )
                            {
                                if ( ValueData.Value[Y + 1, X + 1] )
                                {
                                    //i i i i
                                }
                                else if ( !ValueData.Value[Y + 1, X + 1] )
                                {
                                    //i i i o
                                }
                            }
                            else if ( !ValueData.Value[Y + 1, X] )
                            {
                                if ( ValueData.Value[Y + 1, X + 1] )
                                {
                                    //i i o i
                                }
                                else if ( !ValueData.Value[Y + 1, X + 1] )
                                {
                                    //i i o o
                                }
                            }
                        }
                        else if ( !ValueData.Value[Y, X + 1] )
                        {
                            if ( ValueData.Value[Y + 1, X] )
                            {
                                if ( ValueData.Value[Y + 1, X + 1] )
                                {
                                    //i o i i
                                }
                                else if ( !ValueData.Value[Y + 1, X + 1] )
                                {
                                    //i o i o
                                }
                            }
                            else if ( !ValueData.Value[Y + 1, X] )
                            {
                                if ( ValueData.Value[Y + 1, X + 1] )
                                {
                                    //i o o i
                                    if ( VBMath.Rnd() < 0.5F )
                                    {
                                        ValueData.Value[Y, X] = false;
                                    }
                                    else
                                    {
                                        ValueData.Value[Y + 1, X + 1] = false;
                                    }
                                    Flag = true;
                                }
                                else if ( !ValueData.Value[Y + 1, X + 1] )
                                {
                                    //i o o o
                                }
                            }
                        }
                    }
                    else if ( !ValueData.Value[Y, X] )
                    {
                        if ( ValueData.Value[Y, X + 1] )
                        {
                            if ( ValueData.Value[Y + 1, X] )
                            {
                                if ( ValueData.Value[Y + 1, X + 1] )
                                {
                                    //o i i i
                                }
                                else if ( !ValueData.Value[Y + 1, X + 1] )
                                {
                                    //o i i o
                                    if ( VBMath.Rnd() < 0.5F )
                                    {
                                        ValueData.Value[Y, X + 1] = false;
                                    }
                                    else
                                    {
                                        ValueData.Value[Y + 1, X] = false;
                                    }
                                    Flag = true;
                                }
                            }
                            else if ( !ValueData.Value[Y + 1, X] )
                            {
                                if ( ValueData.Value[Y + 1, X + 1] )
                                {
                                    //o i o i
                                }
                                else if ( !ValueData.Value[Y + 1, X + 1] )
                                {
                                    //o i o o
                                }
                            }
                        }
                        else if ( !ValueData.Value[Y, X + 1] )
                        {
                            if ( ValueData.Value[Y + 1, X] )
                            {
                                if ( ValueData.Value[Y + 1, X + 1] )
                                {
                                    //o o i i
                                }
                                else if ( !ValueData.Value[Y + 1, X + 1] )
                                {
                                    //o o i o
                                }
                            }
                            else if ( !ValueData.Value[Y + 1, X] )
                            {
                                if ( ValueData.Value[Y + 1, X + 1] )
                                {
                                    //o o o i
                                }
                                else if ( !ValueData.Value[Y + 1, X + 1] )
                                {
                                    //o o o o
                                }
                            }
                        }
                    }
                    //when flag, go back one in each direction, incase a new diagonal was created
                    if ( Flag )
                    {
                        if ( X > 0 )
                        {
                            X--;
                        }
                        break;
                    }
                    else
                    {
                        X++;
                    }
                }
                if ( Flag )
                {
                    if ( Y > 0 )
                    {
                        Y--;
                    }
                }
                else
                {
                    Y++;
                }
            }
        }

        public void Expand_One_Tile(clsBooleanMap Source)
        {
            int X = 0;
            int Y = 0;

            SizeCopy(Source);
            for ( Y = 0; Y <= ValueData.Size.Y - 1; Y++ )
            {
                for ( X = 0; X <= Source.ValueData.Size.X - 1; X++ )
                {
                    if ( Source.ValueData.Value[Y, X] )
                    {
                        ValueData.Value[Y, X] = true;
                        if ( Y > 0 )
                        {
                            if ( X > 0 )
                            {
                                ValueData.Value[Y - 1, X - 1] = true;
                            }
                            ValueData.Value[Y - 1, X] = true;
                            if ( X < Source.ValueData.Size.X - 1 )
                            {
                                ValueData.Value[Y - 1, X + 1] = true;
                            }
                        }
                        if ( X > 0 )
                        {
                            ValueData.Value[Y, X - 1] = true;
                        }
                        ValueData.Value[Y, X] = true;
                        if ( X < ValueData.Size.X - 1 )
                        {
                            ValueData.Value[Y, X + 1] = true;
                        }
                        if ( Y < ValueData.Size.Y - 1 )
                        {
                            if ( X > 0 )
                            {
                                ValueData.Value[Y + 1, X - 1] = true;
                            }
                            ValueData.Value[Y + 1, X] = true;
                            if ( X < ValueData.Size.X - 1 )
                            {
                                ValueData.Value[Y + 1, X + 1] = true;
                            }
                        }
                    }
                }
            }
        }

        public void Remove(clsBooleanMap Source, clsBooleanMap Remove)
        {
            int X = 0;
            int Y = 0;

            SizeCopy(Source);
            for ( Y = 0; Y <= Source.ValueData.Size.Y - 1; Y++ )
            {
                for ( X = 0; X <= Source.ValueData.Size.X - 1; X++ )
                {
                    if ( Remove.ValueData.Value[Y, X] )
                    {
                        ValueData.Value[Y, X] = false;
                    }
                    else
                    {
                        ValueData.Value[Y, X] = Source.ValueData.Value[Y, X];
                    }
                }
            }
        }

        public void Combine(clsBooleanMap Source, clsBooleanMap Insert)
        {
            int X = 0;
            int Y = 0;

            SizeCopy(Source);
            for ( Y = 0; Y <= Source.ValueData.Size.Y - 1; Y++ )
            {
                for ( X = 0; X <= Source.ValueData.Size.X - 1; X++ )
                {
                    if ( Insert.ValueData.Value[Y, X] )
                    {
                        ValueData.Value[Y, X] = true;
                    }
                    else
                    {
                        ValueData.Value[Y, X] = Source.ValueData.Value[Y, X];
                    }
                }
            }
        }

        public void Within(clsBooleanMap Interior, clsBooleanMap Exterior)
        {
            int Y = 0;
            int X = 0;
            bool Flag = default(bool);

            SizeCopy(Interior);
            for ( Y = 0; Y <= Interior.ValueData.Size.Y - 1; Y++ )
            {
                for ( X = 0; X <= Interior.ValueData.Size.X - 1; X++ )
                {
                    if ( Interior.ValueData.Value[Y, X] )
                    {
                        Flag = false;
                        if ( Y > 0 )
                        {
                            if ( X > 0 )
                            {
                                if ( !Exterior.ValueData.Value[Y - 1, X - 1] )
                                {
                                    Flag = true;
                                }
                            }
                            if ( !Exterior.ValueData.Value[Y - 1, X] )
                            {
                                Flag = true;
                            }
                            if ( X < Interior.ValueData.Size.X - 1 )
                            {
                                if ( !Exterior.ValueData.Value[Y - 1, X + 1] )
                                {
                                    Flag = true;
                                }
                            }
                        }
                        if ( X > 0 )
                        {
                            if ( !Exterior.ValueData.Value[Y, X - 1] )
                            {
                                Flag = true;
                            }
                        }
                        if ( !Exterior.ValueData.Value[Y, X] )
                        {
                            Flag = true;
                        }
                        if ( X < Interior.ValueData.Size.X - 1 )
                        {
                            if ( !Exterior.ValueData.Value[Y, X + 1] )
                            {
                                Flag = true;
                            }
                        }
                        if ( Y < Interior.ValueData.Size.Y - 1 )
                        {
                            if ( X > 0 )
                            {
                                if ( !Exterior.ValueData.Value[Y + 1, X - 1] )
                                {
                                    Flag = true;
                                }
                            }
                            if ( !Exterior.ValueData.Value[Y + 1, X] )
                            {
                                Flag = true;
                            }
                            if ( X < Interior.ValueData.Size.X - 1 )
                            {
                                if ( !Exterior.ValueData.Value[Y + 1, X + 1] )
                                {
                                    Flag = true;
                                }
                            }
                        }
                        ValueData.Value[Y, X] = !Flag;
                    }
                }
            }
        }
    }
}

File: reference/FlaMEVB/clsBooleanMap.vb
Public Class clsBooleanMap

    Public Class clsValueData
        Public Size As sXY_int
        Public Value(,) As Boolean
    End Class
    Public ValueData As New clsValueData

    Public Sub Blank(SizeX As Integer, SizeY As Integer)

        ValueData.Size.X = SizeX
        ValueData.Size.Y = SizeY
        ReDim ValueData.Value(SizeY - 1, SizeX - 1)
    End Sub

    Public Sub SizeCopy(Source As clsBooleanMap)

        ValueData.Size.X = Source.ValueData.Size.X
        ValueData.Size.Y = Source.ValueData.Size.Y
        ReDim ValueData.Value(ValueData.Size.Y - 1, ValueData.Size.X - 1)
    End Sub

    Public Sub Copy(Source As clsBooleanMap)
        Dim X As Integer
        Dim Y As Integer

        SizeCopy(Source)
        For Y = 0 To Source.ValueData.Size.Y - 1
            For X = 0 To Source.ValueData.Size.X - 1
                ValueData.Value(Y, X) = Source.ValueData.Value(Y, X)
            Next
        Next
    End Sub

    Public Sub Convert_Heightmap(Source As clsHeightmap, AtOrAboveThisHeightEqualsTrue As Long)
        Dim X As Integer
        Dim Y As Integer

        ValueData.Size.X = Source.HeightData.SizeX
        ValueData.Size.Y = Source.HeightData.SizeY
        ReDim ValueData.Value(ValueData.Size.Y - 1, ValueData.Size.X - 1)
        For Y = 0 To Source.HeightData.SizeY - 1
            For X = 0 To Source.HeightData.SizeX - 1
                ValueData.Value(Y, X) = (Source.HeightData.Height(Y, X) >= AtOrAboveThisHeightEqualsTrue)
            Next
        Next
    End Sub

    Public Sub Remove_Diagonals()
        Dim X As Integer
        Dim Y As Integer
        Dim Flag As Boolean

        X = 0
        Y = 0
        Do While Y < ValueData.Size.Y - 1
            X = 0
            Do While X < ValueData.Size.X - 1
                Flag = False
                If ValueData.Value(Y, X) Then
                    If ValueData.Value(Y, X + 1) Then
                        If ValueData.Value(Y + 1, X) Then
                            If ValueData.Value(Y + 1, X + 1) Then
                                'i i i i
                            ElseIf Not ValueData.Value(Y + 1, X + 1) Then
                                'i i i o
                            End If
                        ElseIf Not ValueData.Value(Y + 1, X) Then
                            If ValueData.Value(Y + 1, X + 1) Then
                                'i i o i
                            ElseIf Not ValueData.Value(Y + 1, X + 1) Then
                                'i i o o
                            End If
                        End If
                    ElseIf Not ValueData.Value(Y, X + 1) Then
                        If ValueData.Value(Y + 1, X) Then
                            If ValueData.Value(Y + 1, X + 1) Then
                                'i o i i
                            ElseIf Not ValueData.Value(Y + 1, X + 1) Then
                                'i o i o
                            End If
                        ElseIf Not ValueData.Value(Y + 1, X) Then
                            If ValueData.Value(Y + 1, X + 1) Then
                                'i o o i
                                If Rnd() < 0.5F Then
                                    ValueData.Value(Y, X) = False
                                Else
                                    ValueData.Value(Y + 1, X + 1) = False
                                End If
                                Flag = True
                            ElseIf Not ValueData.Value(Y + 1, X + 1) Then
                                'i o o o
                            End If
                        End If
                    End If
                ElseIf Not ValueData.Value(Y, X) Then
                    If ValueData.Value(Y, X + 1) Then
                        If ValueData.Value(Y + 1, X) Then
                            If ValueData.Value(Y + 1, X + 1) Then
                                'o i i i
                            ElseIf Not ValueData.Value(Y + 1, X + 1) Then
                                'o i i o
                                If Rnd() < 0.5F Then
                                    ValueData.Value(Y, X + 1) = False
                                Else
                                    ValueData.Value(Y + 1, X) = False
                                End If
                                Flag = True
                            End If
                        ElseIf Not ValueData.Value(Y + 1, X) Then
                            If ValueData.Value(Y + 1, X + 1) Then
                                'o i o i
                            ElseIf Not ValueData.Value(Y + 1, X + 1) Then
                                'o i o o
                            End If
                        End If
                    ElseIf Not ValueData.Value(Y, X + 1) Then
                        If ValueData.Value(Y + 1, X) Then
                            If ValueData.Value(Y + 1, X + 1) Then
                                'o o i i
                            ElseIf Not ValueData.Value(Y + 1, X + 1) Then
                                'o o i o
                            End If
                        ElseIf Not ValueData.Value(Y + 1, X) Then
                            If ValueData.Value(Y + 1, X + 1) Then
                                'o o o i
                            ElseIf Not ValueData.Value(Y + 1, X + 1) Then
                                'o o o o
                            End If
                        End If
                    End If
                End If
                'when flag, go back one in each direction, incase a new diagonal was created
                If Flag Then
                    If X > 0 Then X -= 1
                    Exit Do
                Else
                    X += 1
                End If
            Loop
            If Flag Then
                If Y > 0 Then Y -= 1
            Else
                Y += 1
            End If
        Loop
    End Sub

    Public Sub Expand_One_Tile(Source As clsBooleanMap)
        Dim X As Integer
        Dim Y As Integer

        SizeCopy(Source)
        For Y = 0 To ValueData.Size.Y - 1
            For X = 0 To Source.ValueData.Size.X - 1
                If Source.ValueData.Value(Y, X) Then
                    ValueData.Value(Y, X) = True
                    If Y > 0 Then
                        If X > 0 Then
                            ValueData.Value(Y - 1, X - 1) = True
                        End If
                        ValueData.Value(Y - 1, X) = True
                        If X < Source.ValueData.Size.X - 1 Then
                            ValueData.Value(Y - 1, X + 1) = True
                        End If
                    End If
                    If X > 0 Then
                        ValueData.Value(Y, X - 1) = True
                    End If
                    ValueData.Value(Y, X) = True
                    If X < ValueData.Size.X - 1 Then
                        ValueData.Value(Y, X + 1) = True
                    End If
                    If Y < ValueData.Size.Y - 1 Then
                        If X > 0 Then
                            ValueData.Value(Y + 1, X - 1) = True
                        End If
                        ValueData.Value(Y + 1, X) = True
                        If X < ValueData.Size.X - 1 Then
                            ValueData.Value(Y + 1, X + 1) = True
                        End If
                    End If
                End If
            Next
        Next
    End Sub

    Public Sub Remove(Source As clsBooleanMap, Remove As clsBooleanMap)
        Dim X As Integer
        Dim Y As Integer

        SizeCopy(Source)
        For Y = 0 To Source.ValueData.Size.Y - 1
            For X = 0 To Source.ValueData.Size.X - 1
                If Remove.ValueData.Value(Y, X) Then
                    ValueData.Value(Y, X) = False
                Else
                    ValueData.Value(Y, X) = Source.ValueData.Value(Y, X)
                End If
            Next
        Next
    End Sub

    Public Sub Combine(Source As clsBooleanMap, Insert As clsBooleanMap)
        Dim X As Integer
        Dim Y As Integer

        SizeCopy(Source)
        For Y = 0 To Source.ValueData.Size.Y - 1
            For X = 0 To Source.ValueData.Size.X - 1
                If Insert.ValueData.Value(Y, X) Then
                    ValueData.Value(Y, X) = True
                Else
                    ValueData.Value(Y, X) = Source.ValueData.Value(Y, X)
                End If
            Next
        Next
    End Sub

    Public Sub Within(Interior As clsBooleanMap, Exterior As clsBooleanMap)
        Dim Y As Integer
        Dim X As Integer
        Dim Flag As Boolean

        SizeCopy(Interior)
        For Y = 0 To Interior.ValueData.Size.Y - 1
            For X = 0 To Interior.ValueData.Size.X - 1
                If Interior.ValueData.Value(Y, X) Then
                    Flag = False
                    If Y > 0 Then
                        If X > 0 Then
                            If Not Exterior.ValueData.Value(Y - 1, X - 1) Then Flag = True
                        End If
                        If Not Exterior.ValueData.Value(Y - 1, X) Then Flag = True
                        If X < Interior.ValueData.Size.X - 1 Then
                            If Not Exterior.ValueData.Value(Y - 1, X + 1) Then Flag = True
                        End If
                    End If
                    If X > 0 Then
                        If Not Exterior.ValueData.Value(Y, X - 1) Then Flag = True
                    End If
                    If Not Exterior.ValueData.Value(Y, X) Then Flag = True
                    If X < Interior.ValueData.Size.X - 1 Then
                        If Not Exterior.ValueData.Value(Y, X + 1) Then Flag = True
                    End If
                    If Y < Interior.ValueData.Size.Y - 1 Then
                        If X > 0 Then
                            If Not Exterior.ValueData.Value(Y + 1, X - 1) Then Flag = True
                        End If
                        If Not Exterior.ValueData.Value(Y + 1, X) Then Flag = True
                        If X < Interior.ValueData.Size.X - 1 Then
                            If Not Exterior.ValueData.Value(Y + 1, X + 1) Then Flag = True
                        End If
                    End If
                    ValueData.Value(Y, X) = Not Flag
                End If
            Next
        Next
    End Sub
End Class